patches/server/0817-Properly-remove-overriding-potion-effect-modifiers.patch

```diff
@@ -0,0 +1,190 @@
+@@ -10,13 +10,33 @@ public class AbsoptionMobEffect extends MobEffect {
+ 
+     @Override
+     public void removeAttributeModifiers(LivingEntity entity, AttributeMap attributes, int amplifier) {
+-        entity.setAbsorptionAmount(entity.getAbsorptionAmount() - (float)(4 * (amplifier + 1)));
++        entity.setAbsorptionAmount(entity.getAbsorptionAmount() - calculateGrantedAbsorptionHearts(amplifier)); // Paper - properly remove old overridden potion effect
+         super.removeAttributeModifiers(entity, attributes, amplifier);
+     }
+ 
+     @Override
+     public void addAttributeModifiers(LivingEntity entity, AttributeMap attributes, int amplifier) {
+-        entity.setAbsorptionAmount(entity.getAbsorptionAmount() + (float)(4 * (amplifier + 1)));
++        entity.setAbsorptionAmount(entity.getAbsorptionAmount() + calculateGrantedAbsorptionHearts(amplifier)); // Paper - properly remove old overridden potion effect
+         super.addAttributeModifiers(entity, attributes, amplifier);
+     }
++
++    // Paper start - properly remove old overridden potion effect
++    // This method was added notably to
++    //   a) Remove duplication of code
++    //   b) Cause an error on diff if the absorption removal/addition logic above is changed during a minecraft update.
++    public float calculateGrantedAbsorptionHearts(int amplifier) {
++        return 4 * (amplifier + 1);
++    }
++
++    @Override
++    public void updateAttributeModifiers(LivingEntity entity, AttributeMap attributes, int oldAmplifier, int newAmplifier) {
++        super.removeAttributeModifiers(entity, attributes, oldAmplifier);
++        final var currentAbsorptionAmount = entity.getAbsorptionAmount();
++        final var updatedAbsorptionAmount = newAmplifier >= oldAmplifier
++            ? newAmplifier // Always max out absorption if the effect is stronger as this is not a downgrade but a new effect
++            : Math.min(currentAbsorptionAmount, calculateGrantedAbsorptionHearts(newAmplifier)); // Do not replenish lost absorption hearts when moving to weaker potion effect
++        entity.setAbsorptionAmount(updatedAbsorptionAmount);
++        super.addAttributeModifiers(entity, attributes, newAmplifier);
++    }
++    // Paper end - properly remove old overridden potion effect
+ }
+diff --git a/src/main/java/net/minecraft/world/effect/MobEffect.java b/src/main/java/net/minecraft/world/effect/MobEffect.java
+index 8bbb9bdcf95989f1737714655f6f6a269d46d7f2..28a0809a043640d1bd912106d0bf367090252360 100644
+--- a/src/main/java/net/minecraft/world/effect/MobEffect.java
++++ b/src/main/java/net/minecraft/world/effect/MobEffect.java
+@@ -192,6 +192,13 @@ public class MobEffect {
+ 
+     }
+ 
++    // Paper start - properly remove old overridden potion effect - moved here from LivingEntity#onEffectUpdated
++    public void updateAttributeModifiers(final LivingEntity entity, final AttributeMap attributes, final int oldAmplifier, final int newAmplifier) {
```

Comment:
Could you remove the `attributes` parameter? and just get the attributes from the LivingEntity

Reply:
Sure, this method signature mainly followed the existing `removeAttributeModifiers` and `addAttributeModifiers` methods, which both take the `LivingEntity` and its `AttributeMap` as parameters, hence I was inclined to follow in those footsteps (to comply with whatever the reason for the "doubled" passing of the attribute map was on mojangs side).

I'll remove it, thanks for the feedback. Sorry it took so long I just got back from vacation and did not have a development ready machine with me.

Reply:
Removed the passed attribute map.

If this is really that much nicer in terms of code, idk.
E.g. we now need to grab the attribute map explicitly when overwriting the method in the absorption mob effect.